docs/blueprints/EP035.rst

```diff
@@ -0,0 +1,208 @@
+
+Kytos Core Changes
+##################
+
+In order for services to identify which interruptions are affecting the devices
+they use, a mechanism needs to be added to the device representation for
+identify interruption causes.
+
+
+Extend ``status_funcs`` to produce a ``status_reason``
+------------------------------------------------------
+
+Currently, when an object is queried about its ``status``,
+the reason for that status is unknown.
+What we propose here is adding a new property to
+``GenericEntity`` called ``status_reason``,
+and adding in ``status_reason_func``
+to produce values for ``status_reason``.
+The produced ``status_reason`` will be a set of strings,
+so all possible reasons a device might be disable
+can be easily represented.
+
+Example usage:
+
+.. code-block:: python3
+
+    from kytos.core.switch import Switch
+    maintenance_switches = set()
+
+    def switch_maintenance_status(switch):
+        if switch.id in maintenance_switches:
+            return EntityStatus.DOWN
+        return EntityStatus.UP
+
+    Switch.register_status_func('maintenance', switch_maintenance_status)
+
+    def switch_maintenance_reason(switch):
+        if switch.id in maintenance_switches:
+            return {'maintenance'}
+        return set()
+
+    Switch.register_status_reason_func('maintenance', switch_maintenance_reason)
+
+
+New Topology REST Endpoints
+###########################
+
+The following set of new endpoints are intended to allow
+for previewing the affects of adding in interruptions.
+
+
+REST Endpoint ``GET /topology/v3/preview_interruption``
+-------------------------------------------------------
+
+For all know service providers,
+previews affected if the specified interruption occurs now.
+
+Expected Input:
+
+.. code-block:: http
+
+    GET /topology/v3/preview_interruption?switch=<switch_id>&link=<link_id>&interface=<interface_id>&type=<interruption_type>
+
+
+Expected Response:
+
+.. code-block:: json
+
+    {
+        <provider name str>: [<service name str>],
+    }
+
+Side Effects:
+
+- None
+
+Notes:
+
+- Should support multiple key-value pairs with the same key in the query parameters.
+
+
+Endpoint: ``GET <url for previewing>``
+--------------------------------------
+
+For the given service provider,
+previews services affected if the specified interruption occurs now.
+
+Expected Input:
+
+.. code-block:: http
+
+    GET <preview_url>?switch=<switch_id>&link=<link_id>&interface=<interface_id>&type=<interruption_type>
+
+Expected Output:
+
+.. code-block:: json
+
+    [
+        {
+            'id': <service name str>,
+        },
+    ]
+
+Side Effects:
+
+- None
+
+Notes:
+
+- Should support multiple key-value pairs with the same key in the query parameters.
+
+New Kytos Events
+################
+
+The following set of event listeners are intended to be used to notify
+other NApps when a change to the status of a set of devices occurs.
+
+KytosEvent ``topology.interruptions.start``
```

Comment:
Same comment for liveness and link up, later on, but with a lower priority, it might be worth to standardize them to use both ``topology.interruptions.start`` and ``topology.interruptions.end`` too, not sure if it'll be doable in this version, but let's keep on our backlog radar at least.